include/targets/cuda/tunable_kernel.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <tune_quda.h>
+#include <target_device.h>
+#include <kernel_helper.h>
+#include <kernel.h>
+
+#ifdef JITIFY
+#include <jitify_helper2.cuh>
+#endif
+
+namespace quda {
+
+  /**
+     @brief Wrapper around cudaLaunchKernel
+     @param[in] func Device function symbol
+     @param[in] tp TuneParam containing the launch parameters
+     @param[in] arg Host address of argument struct
+     @param[in] stream Stream identifier
+  */
+  qudaError_t qudaLaunchKernel(const void *func, const TuneParam &tp, const qudaStream_t &stream, const void *arg);
+
+  class TunableKernel : public Tunable
+  {
+
+  protected:
+    QudaFieldLocation location;
+
+    virtual unsigned int sharedBytesPerThread() const { return 0; }
+    virtual unsigned int sharedBytesPerBlock(const TuneParam &) const { return 0; }
+
+    template <template <typename> class Functor, bool grid_stride, typename Arg>
+    std::enable_if_t<device::use_kernel_arg<Arg>(), qudaError_t>
+      launch_device(const kernel_t &kernel, const TuneParam &tp, const qudaStream_t &stream, const Arg &arg)
+    {
+#ifdef JITIFY
+      launch_error = launch_jitify<Functor, grid_stride, Arg>(kernel.name, tp, stream, arg);
+#else
+      launch_error = qudaLaunchKernel(kernel.func, tp, stream, static_cast<const void *>(&arg));
+#endif
+      return launch_error;
+    }
+
+    template <template <typename> class Functor, bool grid_stride, typename Arg>
+    std::enable_if_t<!device::use_kernel_arg<Arg>(), qudaError_t>
+      launch_device(const kernel_t &kernel, const TuneParam &tp, const qudaStream_t &stream, const Arg &arg)
+    {
+#ifdef JITIFY
+      launch_error = launch_jitify<Functor, grid_stride, Arg>(kernel.name, tp, stream, arg);
```

Comment:
does the jitify path not use the arg in constant memory?

Reply:
The jitify path can use the arg in constant memory, but we can't do the memcpy here, since the module has not yet been created.  The memcpy is done in the `launch_jitify` function after the module is created (see lib/targets/cuda/jitify_helper.cpp for the details).